src/commands/createWebApp/stacks/getStackPicks.ts

```diff
@@ -0,0 +1,76 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { HttpOperationResponse, ServiceClient } from '@azure/ms-rest-js';
+import { createGenericClient, IAzureQuickPickItem } from 'vscode-azureextensionui';
+import { localize } from '../../../localize';
+import { getWorkspaceSetting } from '../../../vsCodeConfig/settings';
+import { FullJavaStack, FullWebAppStack, IWebAppWizardContext } from '../IWebAppWizardContext';
+import { getJavaLinuxRuntime } from './getJavaLinuxRuntime';
+import { AppStackMinorVersion } from './models/AppStackModel';
+import { JavaContainers, WebAppRuntimes, WebAppStack, WebAppStackValue } from './models/WebAppStackModel';
+
+export async function getStackPicks(context: IWebAppWizardContext): Promise<IAzureQuickPickItem<FullWebAppStack>[]>;
+export async function getStackPicks(context: IWebAppWizardContext, javaVersion: string): Promise<IAzureQuickPickItem<FullJavaStack>[]>;
+export async function getStackPicks(context: IWebAppWizardContext, javaVersion?: string): Promise<IAzureQuickPickItem<FullWebAppStack | FullJavaStack>[]> {
+    const stacks: WebAppStack[] = await getStacks(context);
+    const picks: IAzureQuickPickItem<FullWebAppStack | FullJavaStack>[] = [];
+    for (const stack of stacks) {
+        for (const majorVersion of stack.majorVersions) {
+            const minorVersions: (AppStackMinorVersion<WebAppRuntimes & JavaContainers>)[] = majorVersion.minorVersions
+                // Filter out versions that have major, minor, _and_ patch specified (Aka we only want Node.js 14, not Node.js 14.1.1)
+                .filter(mv => !/\..*\./.test(mv.value))
+                // Filter out versions relative to the specified javaVersion
+                .filter(mv => {
+                    return (!javaVersion && (mv.stackSettings.linuxRuntimeSettings || mv.stackSettings.windowsRuntimeSettings)) ||
+                        (javaVersion && (getJavaLinuxRuntime(javaVersion, mv) || mv.stackSettings.windowsContainerSettings));
+                });
```

Comment:
I found this difficult to understand but maybe that's because I'm not super familiar with these stacks 😅

Reply:
I found a slightly cleaner way to accomplish what I was doing here